Add a cpufreq IOGroup

dannosliwcd opened the pull request:
Related #3258

dannosliwcd committed:
Add an IOGroup for cpufreq

cmcantalupo committed:
[WIP] Introduce SysfsIO interface and empty implementation for CpufreqSysfsIO

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Rename SysfsIO -> SyfsDriver and add static methods for IOGroup factory

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Move CpuinfoIOgroup to SysfsIOGroup

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Rename CpufreqIOGroup to SysfsIOGroup

- The cpufreq specific code still needs to be moved into CpufreqDriver class.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Remove index from property and make index a public enum

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
cleanup plugin method and parse method location

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Implement json parser for modified properties json

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fixup function pointer conversion

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Get rid of public use of driver specific emum

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Simplify SysfsDriver interface

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dannosliwcd committed:
Implement cpufreq sysfs driver

cmcantalupo committed:
[WIP] Refactor SysfsIOGroup to use driver

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dannosliwcd committed:
Fix some of the build errors

dannosliwcd committed:
finish addressing build errors

dannosliwcd committed:
Prefix signals and controls with sysfs iogroup name

dannosliwcd committed:
Clarify description of hardware-provided current frequency

dannosliwcd committed:
apply scaling factors to sysfsiogroup

dannosliwcd committed:
Start working on alias handling

dannosliwcd commented on service/src/CpufreqSysfsDriver.hpp:
```diff
@@ -0,0 +1,53 @@
+            static std::string plugin_name(void);
+            static std::unique_ptr<IOGroup> make_plugin(void);
+        private:
+            enum m_properties_index_e {
```
Can we drop this?

dannosliwcd commented on service/src/CpufreqSysfsDriver.cpp:
```diff
@@ -0,0 +1,180 @@
+
+    int CpufreqSysfsDriver::domain_type(const std::string &) const
+    {
+        return GEOPM_DOMAIN_CPU;
```
Should either make this generically determine what the domain really is, or emit an error when it is not CPU

dannosliwcd commented on service/src/CpufreqSysfsDriver.cpp:
```diff
@@ -0,0 +1,180 @@
+        return oss.str();
+    }
+
+    std::string CpufreqSysfsDriver::control_path(const std::string &control_name,
```
Consider collapsing signal_path and control_path into one function

cmcantalupo committed:
Fixup uses of canonical name and reduce repeated code for error checking

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented on service/src/SysfsIOGroup.cpp:
```diff
@@ -0,0 +1,514 @@
+#include "geopm/Helper.hpp"
+#include "geopm/IOGroup.hpp"
+#include "geopm/PlatformTopo.hpp"
+#include "geopm/json11.hpp"
```
drop...

cmcantalupo committed:
collapse control_path and signal_path to attribute_path method

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Remove properties enum

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Avoid some indentation by throwing sooner

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Move IO_BUFFER_SIZE into SysfsDriver class

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Rename "last_value" to just "value"

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dannosliwcd committed:
Move sysfs json to the docs dir

dannosliwcd committed:
Add a schema for sysfs IOGroup attribute json files

dannosliwcd committed:
Add rst documentation for sysfs IOGroups

- Resolves #3262

dannosliwcd committed:
Fill in missing data in sysfs_attributes_cpufreq.json

cmcantalupo commented on service/src/SysfsIOGroup.hpp:
```diff
@@ -0,0 +1,102 @@
+class SaveControl;
+class IOUring;
+
+// Arbitrary buffer size. We're generally looking at integer values much shorter
+// than 100 digits in length. The IOGroup performs string truncation checks in
+// case that ever changes.
```
Move comment to SysfsDriver.hpp

cmcantalupo committed:
Only load CpufreqSysfsDriver when process has CAP_SYS_ADMIN

- If signal and control pruning were enabled based on calls to access(2)
  then requiring CAP_SYS_ADMIN can be dropped.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Remove duplicate CPU_FREQUENCY_STATUS alias and change max to desired for setspeed

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dannosliwcd committed:
Add a UniqueFd class to close out-of-scope file descriptors

- Related to #3278
- UniqueFd has similar semantics to unique_ptr: Wrapped raw file descriptors
  are accessed through get(), the destructor closes wrapped file
  descriptors, copy operations are disabled, and move operations ensure
  that the move target becomes the new owner for close-on-destruction.

dannosliwcd committed:
Use UniqueFd to manage file descriptors in SysfsIOGroup

- Resolves #3278
- Ensures that opened files in SysfsIOGroup get closed regardless of any
  exceptions encountered.

dannosliwcd committed:
Fix a mis-named function in an error message

dannosliwcd committed:
Prune signals and controls that are not accessible now

- Some signals and controls may be supported by a sysfs IOGroup, but
  might not be accessible to a particular user or at a given point in
  time. Hide signals and controls that are not currently accessible.
- Resolves #3279

dannosliwcd committed:
Start adding unit tests for SysfsIOGroup

dannosliwcd committed:
Add tests for SysfsIOGroup batch IO

dannosliwcd committed:
Add unit tests for CpufreqSysfsDriver

- Related to #3276

cmcantalupo committed:
Fixup EXTRA_DIST targets

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dannosliwcd committed:
Infer cpufreq domain from affected_cpus

- Set the driver's domain to be the smallest topology scope that
  contains all CPUs listed in the affected_cpus attribute
- Related to #3272

cmcantalupo committed:
Move comment to what it is commenting on

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Lower CpufreqIOGroup alias priority

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Whitespace changes for style

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on service/src/IOGroup.cpp:
```diff
@@ -89,6 +90,10 @@ namespace geopm
         // service is not active then loading the ServiceIOGroup will
         // fail.
         if (geopm::has_cap_sys_admin()) {
+            // May want to give this higher priority than the non-safe
+            // msr driver once it is considered more stable.
+            register_plugin(CpufreqSysfsDriver::plugin_name(),
+                            CpufreqSysfsDriver::make_plugin);
```
Was it intentional to move this within the has_cap_sys_admin() check?
The previous version of this had the register_plugin() call *outside* this check.

cmcantalupo replied:
This was a work around for before pruning was implemented.  There is now pruning!

bgeltz commented on service/src/CpufreqSysfsDriver.cpp:
```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2015 - 2022, Intel Corporation
```
```suggestion
 * Copyright (c) 2015 - 2023, Intel Corporation
```

We can update the whole codebase to 2024 later.

bgeltz commented on service/src/CpufreqSysfsDriver.cpp:
```diff
@@ -0,0 +1,188 @@
+#include "geopm/PlatformTopo.hpp"
+#include "geopm_topo.h"
+
+#include <cmath>
+#include <cstring>
+#include <sstream>
```
style: Move up under the c headers.  Look at PlatformIO.cpp as an example.

Desired ordering:
specific class hpp, c headers, c++ headers, GEOPM public headers, other GEOPM headers

dannosliwcd replied:
I've updated my formatter config to sort by:

-1: config.h
0: self header or system under test's header for unit test files
1: C system headers `'<[[:alnum:]._\/]+\.h>' `
2: C++ system headers `'<[[:alnum:].]+>'`
3: Local C headers (generally our public headers. Also including 3rd party such as contrib/gtest/gmock/sqlite): `'"[[:alnum:]._\/]+\.h"'`
4: Local C++ headers (generally our non-public headers. Also including 3rd party): `'"[[:alnum:]._\/]+\.hpp"'`
5: Anything else

Seem okay? Or incorporate 3rd party as part of 1 and 2 or their own sections?

dannosliwcd replied:
Maybe also augment 3 to include both the .h headers and the `geopm/` .hpp headers?

dannosliwcd replied:
Discussed in standup:
Try for 3rd party as the first group after system headers

Yes, use .h and geopm/*.hpp in group 3

bgeltz commented on service/docs/source/geopm_pio_sysfs.7.rst:
```diff
@@ -0,0 +1,105 @@
+hardware signals and controls for sysfs attribute files.
+
+The set of signals and controls supported by an instance of SysfsIOGroup is
+configurable through a json file used by the IOGroup. The json file informs the
```
```suggestion
configurable through a JSON file used by the IOGroup. The json file informs the
```
nit; please make this change everywhere.

bgeltz commented on service/docs/source/geopm_pio_sysfs.7.rst:
```diff
@@ -0,0 +1,105 @@
+it is loaded but in a state that does not expose all of the attributes), or
+may expose additional signals and controls (e.g., if you have configured
+additional instances of SysfsIOGroup that are not bundled with GEOPM). Use
+geopmread and geopmwrite to query the full set of signals and controls that are
```
```suggestion
``geopmread`` and ``geopmwrite`` to query the full set of signals and controls that are
```
nit; please make this change everywhere

cmcantalupo committed:
Move CpufreqSysfs registration outside of cap_sys_admin check

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on service/src/CpufreqSysfsDriver.cpp:
```diff
@@ -0,0 +1,188 @@
+#include <cstring>
+#include <sstream>
+
+// Arbitrary buffer size. We're generally looking at integer values much shorter
+// than 100 digits in length. The IOGroup performs string truncation checks in
+// case that ever changes.
```
This comment seems out of place.

bgeltz commented on service/src/CpufreqSysfsDriver.cpp:
```diff
@@ -0,0 +1,188 @@
+    {
+    }
+
+    CpufreqSysfsDriver::CpufreqSysfsDriver(
+            const PlatformTopo &topo,
+            const std::string &cpufreq_directory)
```
```suggestion
    CpufreqSysfsDriver::CpufreqSysfsDriver(const PlatformTopo &topo,
                                           const std::string &cpufreq_directory)
```

bgeltz commented on service/src/CpufreqSysfsDriver.cpp:
```diff
@@ -0,0 +1,188 @@
+        return m_domain;
+    }
+
+    std::string CpufreqSysfsDriver::attribute_path(
+            const std::string &name,
+            int domain_idx)
```
```suggestion
    std::string CpufreqSysfsDriver::attribute_path(const std::string &name,
                                                   int domain_idx)
```

bgeltz commented on service/src/CpufreqSysfsDriver.cpp:
```diff
@@ -0,0 +1,188 @@
+
+        int cpu_map_fd = open(cpu_map_path.c_str(), O_RDONLY);
+        if (cpu_map_fd == -1) {
+            throw geopm::Exception("SysfsIOGroup failed to open " + cpu_map_path,
```
```suggestion
            throw geopm::Exception("CpufreqSysfsDriver failed to open " + cpu_map_path,
```
I see that this function is called in the constructor of CpufreqSysfsDriver.  Other exception messages from the class start this way.  Unclear if this was intentional or not.  If not, I think we should change all the exceptions in this func to start with "CpufreqSysfsDriver".

bgeltz commented on service/src/SysfsDriver.hpp:
```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2015 - 2022, Intel Corporation
```
```suggestion
 * Copyright (c) 2015 - 2023, Intel Corporation
```
Please check all the license headers for this.

bgeltz commented on service/src/SysfsDriver.hpp:
```diff
@@ -0,0 +1,95 @@
+            /// @brief The properties about a signal or control
+            struct properties_s {
+                std::string name; // The full low level PlatformIO name
+                bool is_writable; // Is control property?
```
Is the comment asking a question we need to bottom out on?

dannosliwcd replied:
That's just the question answered by the value of the bool. I'll rephrase the comment

bgeltz commented on service/src/SysfsIOGroup.cpp:
```diff
@@ -0,0 +1,550 @@
+#include "geopm/Helper.hpp"
+#include "geopm/IOGroup.hpp"
+#include "geopm/PlatformTopo.hpp"
+
```
Headers need to be reorganized ^^

bgeltz commented on service/src/SysfsIOGroup.cpp:
```diff
@@ -0,0 +1,550 @@
+    }
+
+    // Set up mapping between signal and control names and corresponding indices
+    SysfsIOGroup::SysfsIOGroup(
+            std::shared_ptr<SysfsDriver> driver,
+            const PlatformTopo &topo,
+            std::shared_ptr<SaveControl> control_saver,
+            std::shared_ptr<IOUring> batch_reader,
+            std::shared_ptr<IOUring> batch_writer)
```
```suggestion
    SysfsIOGroup::SysfsIOGroup(std::shared_ptr<SysfsDriver> driver,
                               const PlatformTopo &topo,
                               std::shared_ptr<SaveControl> control_saver,
                               std::shared_ptr<IOUring> batch_reader,
                               std::shared_ptr<IOUring> batch_writer)
```

bgeltz commented on service/src/SysfsIOGroup.cpp:
```diff
@@ -0,0 +1,550 @@
+        [signal_name, domain_idx] (const m_pushed_info_s &info) {
+            return info.name == signal_name &&
+                   info.domain_idx == domain_idx;
+
```
```suggestion
```

bgeltz commented on service/src/SysfsIOGroup.cpp:
```diff
@@ -0,0 +1,550 @@
+                }
+                info.buf[bytes_read] = '\0';
+
+                // TODO (dcw): Make sure parser is scaling for us
```
Outstanding?

bgeltz commented on service/src/SysfsIOGroup.cpp:
```diff
@@ -0,0 +1,550 @@
+
+    double SysfsIOGroup::read_signal(const std::string &signal_name, int domain_type, int domain_idx)
+    {
+        std::string cname = check_request(__FUNCTION__, signal_name, "", domain_type, domain_idx);
```
```suggestion
        std::string cname = check_request(__func__, signal_name, "", domain_type, domain_idx);
```
``__FUNCTION__`` is the old way: https://gcc.gnu.org/onlinedocs/gcc/Function-Names.html#:~:text=This%20name%20is%20the%20unadorned,with%20old%20versions%20of%20GCC.
We use ``__func__`` else ware.

cmcantalupo committed:
Adjust SysfsIOGroup to support cases where no controls are available

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dannosliwcd committed:
Make cpufreq domain-scope inference work when multiple policy directories exist

dannosliwcd committed:
Update 2022 years to 2023

cmcantalupo committed:
Check that all required paths exist at SysfsIOGroup construction time

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dannosliwcd committed:
Reorder header includes

dannosliwcd committed:
Captialize references to JSON

dannosliwcd committed:
Show geopmwrite and geopmread in fixed-with font for documentation

dannosliwcd committed:
Remove a cpufreqsysfsdriver comment that has already been replaced by a comment in the header

dannosliwcd committed:
Apply suggestions from code review

Apply indentation changes for function parameters

Co-authored-by: Brad Geltz <brgeltz@gmail.com>

dannosliwcd committed:
Fix wrong class name in CpufreqSysfsDriver error messages

dannosliwcd committed:
Rephrase description of the is_writable sysfs property

dannosliwcd committed:
Update service/src/SysfsIOGroup.cpp

Remove extra empty line

Co-authored-by: Brad Geltz <brgeltz@gmail.com>

dannosliwcd committed:
Remove resolved TODO comments

dannosliwcd committed:
Replace use of __FUNCTION__ with __func__

dannosliwcd committed:
Add a missing include to geopm_topo.h

dannosliwcd committed:
Update header include order

cmcantalupo reviewed (dismissed):
LGTM!

dannosliwcd commented on service/docs/json_data/sysfs_attributes_cpufreq.json:
```diff
@@ -0,0 +1,103 @@
+            "units": "hertz",
+            "writeable": false,
+            "alias": ""
+        },
```
Consider changing the frequency alias to use scaling cur freq which seems to be more widely available. 

Also add some missing CPUINFO_ prefixes on some of the signals.

```suggestion
        "CPUINFO_CUR_FREQ": {
            "attribute": "cpuinfo_cur_freq",
            "scalar": 1e3,
            "description": "The current operating frequency reported by the CPU hardware.",
            "aggregation": "average",
            "behavior": "variable",
            "format": "double",
            "units": "hertz",
            "writeable": false,
            "alias": ""
        },
        "CPUINFO_MAX_FREQ": {
            "attribute": "cpuinfo_max_freq",
            "scalar": 1e3,
            "description": "The maximum allowed frequency to set on the CPU.",
            "aggregation": "average",
            "behavior": "variable",
            "format": "double",
            "units": "hertz",
            "writeable": false,
            "alias": "CPU_FREQUENCY_MAX_AVAIL"
        },
        "CPUINFO_MIN_FREQ": {
            "attribute": "cpuinfo_min_freq",
            "scalar": 1e3,
            "description": "The minimum allowed frequency to set on the CPU.",
            "aggregation": "average",
            "behavior": "variable",
            "format": "double",
            "units": "hertz",
            "writeable": false,
            "alias": "CPU_FREQUENCY_MIN_AVAIL"
        },
        "CPUINFO_TRANSITION_LATENCY": {
            "attribute": "cpuinfo_transition_latency",
            "scalar": 1e-9,
            "description": "The time delay to switch from one P-State to another.",
            "aggregation": "average",
            "behavior": "variable",
            "format": "double",
            "units": "seconds",
            "writeable": false,
            "alias": ""
        },
        "SCALING_CUR_FREQ": {
            "attribute": "scaling_cur_freq",
            "scalar": 1e3,
            "description": "The current requested CPU frequency by the cpufreq scaling driver.",
            "aggregation": "average",
            "behavior": "variable",
            "format": "double",
            "units": "hertz",
            "writeable": false,
            "alias": "CPU_FREQUENCY_STATUS"
        },
```

cmcantalupo committed:
Move the CpufreqDriver plugin load below the ServiceIOGroup

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dannosliwcd committed:
Update names to match cpufreq attribute names

- Also change CPU_FREQUENCY_STATUS to alias CPUFREQ::SCALING_CUR_FREQ

Signed-off-by:  Daniel Wilson <daniel1.wilson@intel.com>

cmcantalupo committed:
Add better exception for bad input and fix unit test with new name

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dannosliwcd committed:
Add alias descriptions for cpufreq signals

dannosliwcd committed:
Check access to specific domain indices when know

cmcantalupo commented on service/src/CpufreqSysfsDriver.cpp:
```diff
@@ -0,0 +1,212 @@
+                return static_cast<double>(NAN);
+            }
+            else {
+                return static_cast<double>(std::stoi(content) * scaling_factor);
```
Add exception handling so that integer conversion results in a NAN in all cases, not just `<unsupported>`.

cmcantalupo committed:
Make any failed conversion result in a NAN when parsing

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add exception handling for std::stoi in constructor

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on service/src/CpufreqSysfsDriver.cpp:
```diff
@@ -0,0 +1,223 @@
+        }
+        double scaling_factor = prop_it->second.scaling_factor;
+        return [scaling_factor](const std::string &content) {
+            double result = static_cast<double>(NAN);
```
```suggestion
            double result = NAN;
```
